### PR TITLE
Add Allocation to SMIR

### PR DESCRIPTION
**Pull request by ouz-a**
As it's discussed [here ](https://rust-lang.zulipchat.com/#narrow/stream/320896-project-stable-mir/topic/Representing.20Constants.20in.20smir)this is an attempt to cover Constants for smir in a different way compared to https://github.com/rust-lang/rust/pull/114342

cc https://github.com/rust-lang/project-stable-mir/issues/15

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_middle/src/mir/interpret/allocation/provenance_map.rs [hunk 1]
```diff
@@ -48,6 +48,10 @@ impl<Prov> ProvenanceMap<Prov> {
     pub fn from_presorted_ptrs(r: Vec<(Size, Prov)>) -> Self {
         ProvenanceMap { ptrs: SortedMap::from_presorted_elements(r), bytes: None }
     }
+
+    pub fn bytes(&self) -> &Option<Box<SortedMap<Size, Prov>>> {
+        &self.bytes
+    }
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/interpret/allocation/provenance_map.rs:54]: we can ignore this, this will always be empty except for miri

File: compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs [hunk 2]
```diff
@@ -185,7 +193,7 @@ impl InitMask {
 // Note: for performance reasons when interning, some of the fields can be partially
 // hashed. (see the `Hash` impl below for more details), so the impl is not derived.
 #[derive(Clone, Debug, Eq, PartialEq, HashStable)]
-struct InitMaskMaterialized {
+pub struct InitMaskMaterialized {
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:196]: keep this private. Instead add an iterator over `u64` that allows getting the blocks out directly (and just produces the same value over and over in case it was lazy). If you need to box those iterators, that's fine.

Comment (oli-obk) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:196]: We can just store a `Vec<u64>` in SMIR, we don't need to use this optimization for now.

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs (hunk 2 above)
Comment (ouz-a) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:196]: > keep this private.

if I make it private I would have to remove `Stable` implementation for it.

Comment (oli-obk) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:196]: you can't, but you can add a function returning the iterator I described to `InitMask`

File: compiler/rustc_middle/src/mir/interpret/allocation.rs [hunk 3]
```diff
@@ -1,7 +1,7 @@
 //! The virtual memory representation of the MIR interpreter.
 
-mod init_mask;
-mod provenance_map;
+pub mod init_mask;
+pub mod provenance_map;
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/allocation.rs:4]: This was all quite deliberately sealed away. Do we really have to make this public? :(

Comment (oli-obk) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:4]: We're gonna make it private before this PR lands. We're not exposing the internals, but adding helper methods that avoid having to expose them.

File: compiler/rustc_smir/src/stable_mir/ty.rs [hunk 4]
```diff
@@ -242,3 +242,35 @@ pub struct BoundTy {
     pub var: usize,
     pub kind: BoundTyKind,
 }
+
+pub type Bytes = Vec<u8>;
+pub type Size = usize;
+pub type Prov = usize;
```
Comment (RalfJung) [inline compiler/rustc_smir/src/stable_mir/ty.rs:248]: This should be `AllocId`. How is that exposed to smir?

File: compiler/rustc_smir/src/stable_mir/ty.rs [hunk 5]
```diff
@@ -242,3 +242,35 @@ pub struct BoundTy {
...
+pub type Align = u64;
+pub type InitMaskMaterialized = Vec<usize>;
+
+#[derive(Clone, Debug)]
+pub struct ProvenanceMap {
+    pub ptrs: Vec<(Size, Prov)>,
+}
```
Comment (RalfJung) [inline compiler/rustc_smir/src/stable_mir/ty.rs:255]: This will not be very useful without doc comments explaining what this is all about.

File: compiler/rustc_middle/src/mir/interpret/allocation.rs (hunk 3 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:4]: You don't need to add any helper methods. The existing methods are sufficient for the interpreter, codegen, and MiniRust.

File: compiler/rustc_smir/src/stable_mir/ty.rs (hunk 4 above)
Comment (oli-obk) [reply compiler/rustc_smir/src/stable_mir/ty.rs:248]: Similar to how we have a table for `DefId`s, we should add something like it for `AllocId`s (instead of exposing the `AllocId`s internals). Since we don't actually need this in this PR yet, we can make `Prov` just be `Opaque` and stick a stringified version of the `AllocId` in it.

File: compiler/rustc_middle/src/mir/interpret/allocation.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:4]: Yes, but these do random access, we want to duplicate the information. Iterating over all bytes is... not great.

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:4]: [Here is how MiniRust does it](https://github.com/RalfJung/minirust/blob/196da9640a4cd608405ad40110d29930827a99a5/tooling/minimize/src/constant.rs#L132-L145)

Comment (oli-obk) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:4]: Yea, I assumed that was too expensive for big allocations, but we can optimize when we see that being an issue

File: compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs (hunk 2 above)
Comment (ouz-a) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:196]: > keep this private. Instead add an iterator over `u64` that allows getting the blocks out directly (and just produces the same value over and over in case it was lazy). If you need to box those iterators, that's fine.

but don't we lose information if it's `InitMaskBlocks::Lazy{true}` or `InitMaskBlocks::Lazy(false)`,  how would someone know if it's lazy or not, since I can't match private value. Maybe something like this ? So people who use smir could identify it it's by checking repeating 0's or 1's
```rust
vec![1;10].iter() // lazy ?
vec![0;10].iter() // not lazy ?
```

Comment (oli-obk) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:196]: Lazy is not information anyone cares about. It is just an internal representation detail. it's a optimization to avoid allocating a vec if all entries are true or all entries are false

Comment (ouz-a) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:196]: Ah I get it now, I didn't know it wasn't useful for smir

File: compiler/rustc_smir/src/stable_mir/ty.rs (hunk 5 above)
Comment (ouz-a) [reply compiler/rustc_smir/src/stable_mir/ty.rs]: I added comments from rustc, I don't know specifics of what smir user would want to know from this. I think it should suffice for now? If not I would love to learn more and add more information/documentation.

File: compiler/rustc_smir/src/stable_mir/ty.rs (hunk 4 above)
Comment (ouz-a) [reply compiler/rustc_smir/src/stable_mir/ty.rs]: I made it `Opaque`

File: compiler/rustc_middle/src/mir/interpret/allocation.rs (hunk 3 above)
Comment (ouz-a) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs]: Made them private, I think in future PR we should def get inspiration from MiniRust.

File: compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs [hunk 6]
```diff
@@ -64,6 +64,13 @@ impl InitMask {
         }
     }
 
+    pub fn get_blocks(&self) -> Box<std::slice::Iter<'_, u64>> {
+        match &self.blocks {
+            InitMaskBlocks::Lazy { state: _ } => Box::new([].iter()),
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:69]: this is wrong, it's not zero sized, but repeating `state` for the length of the entire `bytes` buffer.

But, as Ralf mentioned, we can avoid this entirely by encoding `bytes` as `Vec<Option<u8>>` and using https://github.com/RalfJung/minirust/blob/196da9640a4cd608405ad40110d29930827a99a5/tooling/minimize/src/constant.rs#L132-L145 to generate that `Vec`. Then we don't need an `init_mask` field at all.

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 7]
```diff
@@ -1063,3 +1065,32 @@ impl<'tcx> Stable<'tcx> for rustc_middle::ty::BoundTy {
...
+        let mut blocks = Vec::new();
+        for i in self.init_mask().get_blocks() {
+            blocks.push(*i);
+        }
+        stable_mir::ty::Allocation {
+            bytes: self.get_bytes_unchecked(range).to_vec(),
+            provenance: {
+                let mut ptrs = Vec::new();
+                for (size, prov) in self.provenance().ptrs().iter() {
+                    ptrs.push((size.bytes_usize(), opaque(&prov.0.get())));
```
Comment (oli-obk) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:1085]:
```suggestion
                    ptrs.push((size.bytes_usize(), opaque(prov)));
```

when we render it opaquely, just render the full thing instead of its internal information

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 8]
```diff
@@ -1063,3 +1063,38 @@ impl<'tcx> Stable<'tcx> for rustc_middle::ty::BoundTy {
...
+
+impl<'tcx> Stable<'tcx> for mir::interpret::Allocation {
+    type T = stable_mir::ty::Allocation;
+
+    fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
+        let size = self.size();
+        let mut blocks = Vec::new();
+        for i in self.init_mask().get_blocks() {
+            blocks.push(*i);
+        }
```
Comment (oli-obk) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:1075]: This and `get_blocks` is now dead code

File: compiler/rustc_smir/src/stable_mir/ty.rs [hunk 9]
```diff
@@ -242,3 +242,36 @@ pub struct BoundTy {
...
+#[derive(Clone, Debug)]
+pub enum InitMaskBlocks {
+    Materialized(InitMaskMaterialized),
+}
+
+#[derive(Clone, Debug)]
+pub struct InitMask {
+    pub blocks: InitMaskBlocks,
+    pub len: Size,
+}
```
Comment (oli-obk) [inline compiler/rustc_smir/src/stable_mir/ty.rs:269]: THis isn't needed anymore either

**Comment by oli-obk** [comment]
Please squash the commits after all comments have been addressed

File: compiler/rustc_middle/src/mir/interpret/allocation.rs [hunk 10]
```diff
@@ -17,14 +17,14 @@ use rustc_data_structures::intern::Interned;
...
 
+use self::provenance_map::{ProvenanceCopy, ProvenanceMap};
 use super::{
     read_target_uint, write_target_uint, AllocId, InterpError, InterpResult, Pointer, Provenance,
     ResourceExhaustionInfo, Scalar, ScalarSizeMismatch, UndefinedBehaviorInfo, UninitBytesAccess,
     UnsupportedOpInfo,
 };
 use crate::ty;
 use init_mask::*;
-use provenance_map::*;
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/allocation.rs:27]: Looks like some spurious remainders, these are now just unnecessary changes.

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 11]
```diff
@@ -1063,3 +1063,34 @@ impl<'tcx> Stable<'tcx> for rustc_middle::ty::BoundTy {
...
+
+    fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
+        let size = self.size();
+        let mut bytes: Vec<Option<u8>> = self
+            .inspect_with_uninit_and_ptr_outside_interpreter(0..size.bytes_usize())
+            .iter()
+            .copied()
+            .map(Some)
+            .collect();
+        for (i, b) in bytes.iter_mut().enumerate() {
```
Comment (RalfJung) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:1078]: FWIW there is a ore efficient way of doing this with `range_as_init_chunks`. Not sure if it matters though.

Comment (ouz-a) [reply compiler/rustc_smir/src/rustc_smir/mod.rs:1078]: I think we will focus more on performance as a whole after doing benchmarks and whatnot
